src/codegen/compile.ml

```diff
@@ -4164,7 +4164,81 @@ module RTS_Exports = struct
 
 end (* RTS_Exports *)
 
-module Serialization = struct
+module type Stream = sig
+  (*               env    ref    code *)
+  val write_byte : E.t -> G.t -> G.t -> G.t
```

Comment:
It would be good to note which arguments are guaranteed to be used once (the `code`), and which ones these function like to duplicate and should ideally be a simple `get_foo` (the `ref`).

At some point we might refactor this file to let the type system distinguish the two. Maybe.

Reply:
Yes, this needs some care in formulation. `ref` is the moral equivalent of `this` in C++. I could have
``` camp
module type Stream = sig
  type t
  val write_byte : E.t -> G.t -> t -> G.t
  ...
```
Looks more idiomatic, and in the concretisation we would always have `type t = G.t`. What do you think?

Reply:
Hmm, that might be hiding too much implementation - the caller still needs to be aware that the `this` is a piece of wasm code that only makes sense within a certain function, right?

Reply:
Okay, I don't see an obvious way forward, so I am being lazy and just document the facts: see 13cc22c.